gqltest: add docs to run and add tests (#12285)
sourcegraph/sourcegraph: dev/gqltest/README.md

# and copy relevant credentials to here.
export AWS_ACCESS_KEY_ID=<REDACTED>
export AWS_SECRET_ACCESS_KEY=<REDACTED>
export AWS_CODE_COMMIT_USERNAME=<REDACTED>
export AWS_CODE_COMMIT_PASSWORD=<REDACTED>

export BITBUCKET_SERVER_URL=<REDACTED>
export BITBUCKET_SERVER_TOKEN=<REDACTED>
export BITBUCKET_SERVER_USERNAME=<REDACTED>
```

You need to run `direnv allow` after editing the `.envrc` file.

## How to run tests

GraphQL-based integration tests are running against a live Sourcegraph instance, the eaiset way to make one is by booting up a single Docker container:

```sh
# For easier testing, run Sourcegraph instance without volume,
# so it always starts from a clean state.
docker run --publish 7080:7080 --rm sourcegraph/server:insiders
```

Once the the instance is live (look for the log line `✱ Sourcegraph is ready at: http://127.0.0.1:7080`), you can open another terminal tab to run these tests under this directory (`dev/gqltest`):

```sh
→ go test -tags "gqltest"
2020/07/17 14:17:32 Site admin has been created: gqltest-admin
PASS
ok  	github.com/sourcegraph/sourcegraph/dev/gqltest	31.521s
```

> Generally, you're able to repeatly running these tests regardless of any failures because tests are written in the way that cleans up and restores to the previous state. It is aware of if the instance has been initialized, so you can focus on debugging tests.

Reviewer: ```suggestion
Generally, you're able to repeatedly run these tests regardless of any failures because tests are written in the way that cleans up and restores to the previous state. It is aware of if the instance has been initialized, so you can focus on debugging tests.
```

Also I'm not sure I follow the part rerunning tests--it seems to say "if a test fails, that's OK, you can rerun it" but why wouldn't that be the case normally? I think I'm missing some detail about how it works that seems important, maybe?

Author: Hmm, good point, I think my intention was to "you don't have to kill and restart the instance, you can just edit your tests and rerun them". Does that help you understand?

Reviewer: That's clear. Though it would also be surprising if I did have to restart the instance (i.e., my default assumption is it isn't needed).

Because we're using the standard Go test framework, you are able to just run a single or subset of these tests:

```sh
→ go test -tags "gqltest" -run TestSearch
2020/07/17 14:20:59 Site admin authenticated: gqltest-admin
PASS
ok  	github.com/sourcegraph/sourcegraph/dev/gqltest	3.073s
```

### Testing against local dev instance

It is not required to boot up a single Docker container to run these tests, which means it's also possible to run these tests against any Sourcegraph instance, for example, your local dev instance:

```sh
go test -tags "gqltest" -base-url "http://localhost:3080" -email "joe@sourcegraph.com" -username "joe" -password "<REDACTED>"

> ```

Reviewer: I was wondering this after reading the docker image part. I think it would flow better to put before the "Generally, you're able to repeat..." sentence.

Author: Sounds good, let me move it up.

## How to add new tests

Adding new tests to this test suite is as easy as adding a Go test, here are some general rules to follow:

- Add `// +build gqltest` on every new file as the very first line.

> - Use `gqltest-` prefix for entity name, and be as specific be as possible for eaiser debugging, e.g. `gqltest-org-user-1`.

Reviewer: ```suggestion
- Use `gqltest-` prefix for entity name, and be as specific as possible for easier debugging, e.g. `gqltest-org-user-1`.
```

> - Restore to the previous state regardless of failures, including:

Reviewer: ```suggestion
- Restore the previous state regardless of failures, including:
```
